### DISCUSSION TITLE
Add ability to customize span names for TraceLayer

### DISCUSSION THREAD

**Discussion by horseinthesky** [tower-rs/tower-http#470] [Q&A] [answered]
- [ x] I have looked for existing issues (including closed) about this

## Feature Request
At the moment addition of `TraceLayer` provides you no ability to name your span
```
let app = Router::new()
    .route("/api/devices", get(devices))
    .layer(
        TraceLayer::new_for_http()
            .make_span_with(DefaultMakeSpan::new().include_headers(true)),
    )
    .route(
        "/metrics",
        get(|| async { "metrics" }),
    );
```

As the result all spans are named `request`.

### Motivation
It would be nice see method+endpoint uri instead. Something like `GET /api/devices`.
This would simplify working with spans on a collector side.

### Proposal
I'm very new to Rust. If I'm not mistaken this is the place where `request` name code from
https://github.com/tower-rs/tower-http/blob/main/tower-http/src/trace/make_span.rs#L84-L101

Not sure how to achieve this.

**Comment by jplatte** [comment]
This is already possible, using `make_span_with`. See for example the following snippet in the docs: https://docs.rs/tower-http/latest/tower_http/trace/index.html#recording-fields-on-the-span

**Comment by horseinthesky** [comment]
> This is already possible, using `make_span_with`. See for example the following snippet in the docs: [docs.rs/tower-http/latest/tower_http/trace/index.html#recording-fields-on-the-span](https://docs.rs/tower-http/latest/tower_http/trace/index.html#recording-fields-on-the-span)

Thanks a lot.

Every other attribute works great but struggling setting up the name
![image](https://github.com/tower-rs/tower-http/assets/19948672/1eb77a6d-35dc-4298-afb1-555bb0268f19)

**Reply by jplatte** [reply]
Oh, you're talking about a runtime-specified span name. That's not really possible with tracing. You can look for / open an issue [on its issue tracker](https://github.com/tokio-rs/tracing/issues), but honestly I don't expect them to _ever_ support that. It's possible to hack around that like I previously did [here](https://github.com/matrix-org/matrix-rust-sdk/blob/9228ad2f5948db120a826261f236dfd68ce0e8f2/bindings/matrix-sdk-ffi/src/tracing.rs#L97), but that could possibly break at any time when you upgrade tracing.

**Reply by horseinthesky** [reply]
@jplatte I've found a crate that actully does this but I dont' have enough Rust knowledge to understand how =(
https://github.com/davidB/tracing-opentelemetry-instrumentation-sdk/tree/main/axum-tracing-opentelemetry

**Reply by jplatte** [reply]
Where are you seeing dynamic names in that crate?

**Reply by horseinthesky** [reply]
Here is an example:
```
use axum::{routing::get, Json, Router};
use opentelemetry::KeyValue;
use axum_tracing_opentelemetry::middleware::OtelAxumLayer;
use opentelemetry_otlp::WithExportConfig;
use opentelemetry_sdk::Resource;
use opentelemetry_semantic_conventions as semconv;
use serde::Serialize;
use tracing::{debug, info, info_span, warn};
use tracing_subscriber::prelude::*;

mod devices;
use devices::{get_devices, Device};

#[derive(Serialize)]
struct Response<'a> {
    status: &'a str,
    message: &'a str,
}

fn init_otlp() {
    // Create OTLP gRPC exporter
    let exporter = opentelemetry_otlp::new_exporter()
        .grpcio()
        .with_endpoint("localhost:4317");

    // Create a resource
    let resource = Resource::new([KeyValue::new(
        semconv::resource::SERVICE_NAME,
        "rust-app",
    )]);

    // Create tracer
    let tracer = opentelemetry_otlp::new_pipeline()
        .tracing()
        .with_exporter(exporter)
        .with_trace_config(
            opentelemetry_sdk::trace::config().with_resource(resource),
        )
        .install_batch(opentelemetry_sdk::runtime::Tokio)
        .expect("should create a tracer");

    // Create an opentelemetry layer
    let otlp_layer = tracing_opentelemetry::layer().with_tracer(tracer);

    // Create a subscriber
    let subscriber = tracing_subscriber::Registry::default().with(otlp_layer);

    // Set the global subscriber for the app
    tracing::subscriber::set_global_default(subscriber).unwrap();
}

#[tokio::main]
async fn main() -> Result<(), axum::BoxError> {
    init_otlp();

    let app = Router::new()
        .route("/api/devices", get(devices))
        .layer(OtelAxumLayer::default())
        .route("/health", get(health));

    let listener = tokio::net::TcpListener::bind("127.0.0.1:8000")
        .await
        .unwrap();

    axum::serve(listener, app)
        .await?;

    Ok(())
}

async fn health() -> Json<Response<'static>> {
    Json(Response {
        status: "ok",
        message: "up",
    })
}

#[tracing::instrument]
async fn devices() -> Json<Vec<Device<'static>>> {
    info_span!("internal").in_scope(|| {
        warn!("do stuff inside internal");
    });

    Json(get_devices())
}
```

`get_devices()` just returns some static content. Doens't really matter.

Here is how traces look like in Tempo
![image](https://github.com/tower-rs/tower-http/assets/19948672/55bd8d23-2975-4826-a0e2-e076eea5024f)

Here is full trace info
![image](https://github.com/tower-rs/tower-http/assets/19948672/fa77945b-48bb-4f06-be7a-f1cc9daac134)

So as I've said it's able to put HTTP method and URL to the trace name whatever they are i.e. dynamycally.

**Reply by jplatte** [reply]
Sure, there is a name field there that has a dynamic value. I don't think it's the same as the tracing span name though.

**Reply by horseinthesky** [reply]
What do you mean by saying "isn't the same"?

If I replace
```
.layer(OtelInResponseLayer::default())
```
with
```
.layer(
    TraceLayer::new_for_http()
        .make_span_with(DefaultMakeSpan::new().include_headers(true)),
)
```
there is no more method/url info in traces names.
![image](https://github.com/tower-rs/tower-http/assets/19948672/51b45ee7-4940-44cc-b772-61c39a110d5e)
detail
![image](https://github.com/tower-rs/tower-http/assets/19948672/37c09348-8927-4000-8c8b-cc1082d12e94)

**Reply by jplatte** [reply, marked as answer]
I mean that the `GET /api/devices` string from your previous screenshot is not the tracing span's [`.metadata()?.name()`](https://docs.rs/tracing/latest/tracing/struct.Metadata.html#method.name). And indeed I think I found how it works: If there is a field named [`otel.name`](https://github.com/davidB/tracing-opentelemetry-instrumentation-sdk/blob/f000cd83cf4fe2c8c7db97f907fc5b124a2e4418/axum-tracing-opentelemetry/src/middleware/trace_extractor.rs#L129) on the span, that's used in the `Name` column. Try writing something like this for `make_span_with`:

```rust
let name = format!("{whatever} {you} {want}");
info_span!(
    "request",
    otel.name = name,
    extra_field = value,
)
```

**Reply by horseinthesky** [reply]
Awesome! =)

Solution
```
.layer(
    TraceLayer::new_for_http().make_span_with(
        |request: &Request<_>| {
            let name =
                format!("{} {}", request.method(), request.uri());

            tracing::debug_span!(
                "request",
                otel.name = name,
                method = %request.method(),
                uri = %request.uri(),
                headers = ?request.headers(),
                version = ?request.version(),
            )
        },
    ),
)
```

Traces
![image](https://github.com/tower-rs/tower-http/assets/19948672/2dcbea82-1711-4ba1-bf8a-b3e58491f0e4)

Details
![image](https://github.com/tower-rs/tower-http/assets/19948672/71723043-412f-4394-bba1-1b83fab26304)

Thank you
